src/Bicep.Local.Extension/ProviderExtension.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime;
+using Microsoft.AspNetCore.Builder;
+using Bicep.Local.Extension.Rpc;
+using CommandLine;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Bicep.Local.Extension.Protocol;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+
+namespace Bicep.Local.Extension;
+
+public class ProviderExtension
+{
+    internal class CommandLineOptions
+    {
+        [Option("socket", Required = false, HelpText = "The path to the domain socket to connect on")]
+        public string? Socket { get; set; }
+
+        [Option("wait-for-debugger", Required = false, HelpText = "If set, wait for a dotnet debugger to be attached before starting the server")]
+        public bool WaitForDebugger { get; set; }
+    }
+
+    public static async Task Run(Action<ResourceDispatcherBuilder> registerHandlers, string[] args)
+        => await RunWithCancellationAsync(async cancellationToken =>
+        {
+            if (IsTracingEnabled)
+            {
+                Trace.Listeners.Add(new TextWriterTraceListener(Console.Error));
+            }
+
+            var extension = new ProviderExtension();
+
+            await extension.RunAsync(args, registerHandlers, cancellationToken);
+        });
+
+    public async Task RunAsync(string[] args, Action<ResourceDispatcherBuilder> registerHandlers, CancellationToken cancellationToken)
+    {
+        var parser = new Parser(settings =>
+        {
+            settings.IgnoreUnknownArguments = true;
+        });
+
+        await parser.ParseArguments<CommandLineOptions>(args)
+            .WithNotParsed((x) => System.Environment.Exit(1))
+            .WithParsedAsync(async options => await RunServer(registerHandlers, options, cancellationToken));
+    }
+
+    private static async Task RunServer(Action<ResourceDispatcherBuilder> registerHandlers, CommandLineOptions options, CancellationToken cancellationToken)
+    {
+        if (options.WaitForDebugger)
+        {
+            // exit if we don't have a debugger attached within 5 minutes
+            var debuggerTimeoutToken = CancellationTokenSource.CreateLinkedTokenSource(
+                cancellationToken,
+                new CancellationTokenSource(TimeSpan.FromMinutes(5)).Token).Token;
```

Comment:
Nit: 5 minutes seems like a long time. Should this ever take more than a few seconds?

Reply:
Yes - this is a manual process. The extension waits in a "sleep" mode, while the extension author has to open VS/VSCode, and choose "Attach to process".